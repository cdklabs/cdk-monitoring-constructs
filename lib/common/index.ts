export * from "./alarm";
export * from "./metric";
export * from "./monitoring";
export * from "./url";
export * from "./widget";
export * from "./strings";
