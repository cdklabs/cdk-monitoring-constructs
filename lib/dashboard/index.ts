export * from "./BitmapDashboard";
export * from "./DashboardRenderingPreference";
export * from "./DynamicDashboardSegment";
export * from "./DashboardSegment";
export * from "./DashboardWithBitmapCopy";
export * from "./DefaultDashboardFactory";
export * from "./DynamicDashboardFactory";
export * from "./DefaultWidgetFactory";
export * from "./IDashboardFactory";
export * from "./IDynamicDashboardFactory";
export * from "./IWidgetFactory";
export * from "./MonitoringNamingStrategy";
export * from "./SingleWidgetDashboardSegment";
export * from "./widget";
