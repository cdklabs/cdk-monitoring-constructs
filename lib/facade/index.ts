export * from "./IMonitoringAspect";
export * from "./MonitoringFacade";
export * from "./MonitoringAspect";
