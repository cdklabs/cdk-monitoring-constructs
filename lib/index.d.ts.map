{"version":3,"file":"index.d.ts","sourceRoot":"","sources":["index.ts"],"names":[],"mappings":"AAAA,cAAc,UAAU,CAAC;AACzB,cAAc,aAAa,CAAC;AAC5B,cAAc,UAAU,CAAC;AACzB,cAAc,cAAc,CAAC"}