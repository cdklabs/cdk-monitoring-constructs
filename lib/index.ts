export * from "./common";
export * from "./dashboard";
export * from "./facade";
export * from "./monitoring";
