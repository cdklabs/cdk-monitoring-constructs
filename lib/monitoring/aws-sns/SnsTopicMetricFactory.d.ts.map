{"version":3,"file":"SnsTopicMetricFactory.d.ts","sourceRoot":"","sources":["SnsTopicMetricFactory.ts"],"names":[],"mappings":"AAAA,OAAO,EAAE,MAAM,EAAE,MAAM,qBAAqB,CAAC;AAE7C,OAAO,EAAE,aAAa,EAAE,MAAM,cAAc,CAAC;AAE7C,MAAM,WAAW,0BAA0B;IACzC,QAAQ,CAAC,KAAK,EAAE,MAAM,CAAC;CACxB;AAED,qBAAa,qBAAqB;IAChC,SAAS,CAAC,QAAQ,CAAC,aAAa,EAAE,aAAa,CAAC;IAChD,SAAS,CAAC,QAAQ,CAAC,KAAK,EAAE,MAAM,CAAC;gBAErB,aAAa,EAAE,aAAa,EAAE,KAAK,EAAE,0BAA0B;IAK3E,0BAA0B;IAQ1B,0BAA0B;IAQ1B,+BAA+B;IAQ/B,iCAAiC;CAOlC"}