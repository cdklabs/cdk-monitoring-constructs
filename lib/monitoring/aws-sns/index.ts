export * from "./SnsTopicMetricFactory";
export * from "./SnsTopicMonitoring";
