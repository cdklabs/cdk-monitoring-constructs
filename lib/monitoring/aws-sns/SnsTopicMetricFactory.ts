import { ITopic } from "aws-cdk-lib/aws-sns";

import {
  BaseMetricFactory,
  BaseMetricFactoryProps,
  MetricFactory,
} from "../../common";

export interface SnsTopicMetricFactoryProps extends BaseMetricFactoryProps {
  readonly topic: ITopic;
}

export class SnsTopicMetricFactory extends BaseMetricFactory<SnsTopicMetricFactoryProps> {
  protected readonly topic: ITopic;

  constructor(metricFactory: MetricFactory, props: SnsTopicMetricFactoryProps) {
    super(metricFactory, props);

    this.topic = props.topic;
  }

  metricIncomingMessageCount() {
    return this.metricFactory.adaptMetric(
      this.topic.metricNumberOfMessagesPublished({
        label: "Incoming",
        region: this.region,
        account: this.account,
      }),
    );
  }

  metricOutgoingMessageCount() {
    return this.metricFactory.adaptMetric(
      this.topic.metricNumberOfNotificationsDelivered({
        label: "Outgoing",
        region: this.region,
        account: this.account,
      }),
    );
  }

  metricAverageMessageSizeInBytes() {
    return this.metricFactory.adaptMetric(
      this.topic.metricPublishSize({
        label: "Size",
        region: this.region,
        account: this.account,
      }),
    );
  }

  metricNumberOfNotificationsFailed() {
    return this.metricFactory.adaptMetric(
      this.topic.metricNumberOfNotificationsFailed({
        label: "Failed",
        region: this.region,
        account: this.account,
      }),
    );
  }
}
