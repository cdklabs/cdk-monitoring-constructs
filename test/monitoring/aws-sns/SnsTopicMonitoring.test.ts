import { Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { Topic } from "aws-cdk-lib/aws-sns";

import { AlarmWithAnnotation, SnsTopicMonitoring } from "../../../lib";
import { TestMonitoringScope } from "../TestMonitoringScope";

test("snapshot test: no alarms", () => {
  const stack = new Stack();

  const scope = new TestMonitoringScope(stack, "Scope");

  const topic = new Topic(stack, "Topic", {
    topicName: "DummyTopic",
  });

  new SnsTopicMonitoring(scope, {
    topic,
  });

  // alternative: use reference

  new SnsTopicMonitoring(scope, {
    topic: Topic.fromTopicArn(
      stack,
      "DummyTopicRef",
      "arn:aws:sns:us-east-2:123456789012:DummyTopicRef",
    ),
  });

  expect(Template.fromStack(stack)).toMatchSnapshot();
});

test("snapshot test: all alarms", () => {
  const stack = new Stack();

  const scope = new TestMonitoringScope(stack, "Scope");

  const topic = new Topic(stack, "Topic", {
    topicName: "DummyTopic",
  });

  let numAlarmsCreated = 0;

  new SnsTopicMonitoring(scope, {
    topic,
  });

  new SnsTopicMonitoring(scope, {
    topic,
    addMessageNotificationsFailedAlarm: {
      Warning: {
        maxNotificationsFailedCount: 5,
      },
    },
    addMinNumberOfMessagesPublishedAlarm: {
      Warning: {
        minMessagesPublishedCount: 100,
      },
    },
    addMaxNumberOfMessagesPublishedAlarm: {
      Warning: {
        maxMessagesPublishedCount: 200,
      },
    },
    useCreatedAlarms: {
      consume(alarms: AlarmWithAnnotation[]) {
        numAlarmsCreated = alarms.length;
      },
    },
  });

  expect(numAlarmsCreated).toStrictEqual(3);
  expect(Template.fromStack(stack)).toMatchSnapshot();
});
