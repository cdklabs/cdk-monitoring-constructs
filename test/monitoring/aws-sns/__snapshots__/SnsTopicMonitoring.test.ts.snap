// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test: all alarms 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "ScopeTestDummyTopicTopicNotificationsFailedWarning39FB74A9": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Number of failed sns messages exceeded threshold of 5.",
        "AlarmName": "Test-DummyTopic-Topic-Notifications-Failed-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Failed",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "TopicName",
                    "Value": Object {
                      "Fn::GetAtt": Array [
                        "TopicBFC7AF6E",
                        "TopicName",
                      ],
                    },
                  },
                ],
                "MetricName": "NumberOfNotificationsFailed",
                "Namespace": "AWS/SNS",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 5,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyTopicTopicPublishedHighWarningCDE28531": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Number of SNS messages published is too high.",
        "AlarmName": "Test-DummyTopic-Topic-Published-High-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Incoming",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "TopicName",
                    "Value": Object {
                      "Fn::GetAtt": Array [
                        "TopicBFC7AF6E",
                        "TopicName",
                      ],
                    },
                  },
                ],
                "MetricName": "NumberOfMessagesPublished",
                "Namespace": "AWS/SNS",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 200,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyTopicTopicPublishedLowWarning53240D68": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Number of SNS messages published is too low.",
        "AlarmName": "Test-DummyTopic-Topic-Published-Low-Warning",
        "ComparisonOperator": "LessThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Incoming",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "TopicName",
                    "Value": Object {
                      "Fn::GetAtt": Array [
                        "TopicBFC7AF6E",
                        "TopicName",
                      ],
                    },
                  },
                ],
                "MetricName": "NumberOfMessagesPublished",
                "Namespace": "AWS/SNS",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 100,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TopicBFC7AF6E": Object {
      "Properties": Object {
        "TopicName": "DummyTopic",
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`snapshot test: no alarms 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "TopicBFC7AF6E": Object {
      "Properties": Object {
        "TopicName": "DummyTopic",
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
